packages/clay-core/src/tree-view/useTree.ts

```diff
@@ -265,6 +310,38 @@ export function createImmutableTree<T extends Array<Record<string, any>>>(
 
 					break;
 				}
+
+				case 'remove': {
+					const {path} = patch;
+
+					const nodeToRemove = nodeByPath(path);
+
+					if (nodeToRemove.parent) {
+						nodeToRemove.parent[nestedKey] = nodeToRemove.parent[
+							nestedKey
+						].filter(
+							(_item: any, index: number) =>
+								index !== nodeToRemove.index
+						);
+					} else {
+						immutableTree = immutableTree.filter(
+							(_item: any, index: number) =>
+								index !== nodeToRemove.index
+						) as T;
+					}
+
+					break;
+				}
+
+				case 'rename': {
+					const {name, path} = patch;
+
+					const nodeToRename = nodeByPath(path);
+					nodeToRename.item.name = name;
```

Comment:
I think we can say that `onRenameItem` should return the object instead of changing a property of the object or we would have to make it configurable to remain data agnostic.

Reply:
@matuzalemsteles while trying this out I saw a problem: We can't pass the entire item to the `onRenameItem` callback because it would allow modifying important "hidden" properties like `index`, `indexes` or `key`, but at the same time if we omit passing those properties the operation "fails" when trying to rename a item in the "top" level because we loose that information. Since this was hard to solve, I'm actually passing the entire item in the callback, you can see this in https://github.com/liferay/clay/pull/4348/commits/570dfd501326e68d455902864a9460c58de29b9a

Since it's hard to explain, we can talk about it via Slack if you think it's necessary.

Reply:
@matuzalemsteles just to be more clear, I'm not sure it's should be the TreeView's responsibility

Reply:
I tried a bunch of different things, but not that :smile: Thanks for the tip